src/EFCore.PG/Query/ExpressionVisitors/NpgsqlSqlTranslatingExpressionVisitor.cs

```diff
@@ -64,29 +38,24 @@ public class NpgsqlSqlTranslatingExpressionVisitor : SqlTranslatingExpressionVis
             : base(dependencies, queryModelVisitor, targetSelectExpression, topLevelPredicate, inProjection)
             => _queryModelVisitor = queryModelVisitor;
 
-        /// <inheritdoc />
-        protected override Expression VisitSubQuery(SubQueryExpression expression)
-            => base.VisitSubQuery(expression) ?? VisitLikeAnyAll(expression) ?? VisitEqualsAny(expression);
+        #region Overrides
 
         /// <inheritdoc />
         protected override Expression VisitBinary(BinaryExpression expression)
         {
-            if (expression.NodeType == ExpressionType.ArrayIndex)
+            var left = Visit(expression.Left);
```

Comment:
The point of the previous code was to immediately delegate to `base.VisitBinary()` in all cases, *unless* when the expression's node type is `ArrayIndex`. In the new code you call `Visit()` on Left and Right, and then it gets called again if the expression isn't `ArrayIndex` or `Index`. This could be bad for performance and may potentially have other side-effects - when we're changing the query pipeline I'd rather be as little invasive as we can.

So I'd start with a switch on the NodeType, which immediately returns `base.VisitBinary()` for anything but what we handle here.

Reply:
I've reworked this into a switch to cut down on the possibility of wasted visits. How does it look now?